Cue/app/tabs/discover/DiscoverHome.js

```diff
@@ -25,37 +28,60 @@ const styles = {
 
 type Props = {
   navigator: Navigator,
-  onPressMenu?: () => void
+  onPressMenu?: () => void,
+
+  newDecks: ?Array<DeckMetadata>,
+  topDecks: ?Array<DeckMetadata>,
+  onRefreshDiscover: () => any,
 }
 
 class DiscoverHome extends React.Component {
   props: Props
 
   state: {
-    dataSource: ListView.DataSource
+    dataSource: ListView.DataSource,
+    refreshing: boolean
   }
 
+  ds: ListView.DataSource
+
   constructor(props) {
     super(props)
 
-    let ds = new ListView.DataSource({
+    this.ds = new ListView.DataSource({
       rowHasChanged: (r1, r2) => r1 !== r2,
       sectionHeaderHasChanged: (s1, s2) => s1 !== s2
     })
 
-    Promise.all([DiscoverApi.fetchTop(), DiscoverApi.fetchNew()])
-      .then((decks) => {
-        this.setState({
-          ...this.state,
-          dataSource: ds.cloneWithRowsAndSections({
-            'Top Decks': [decks[0]],
-            'New Decks': [decks[1]],
-          })
-        })
-      })
-
-    this.state = {
-      dataSource: ds.cloneWithRowsAndSections({})
+    this.state = this._getNewState(props)
+  }
+
+  componentWillReceiveProps(newProps) {
+    this.setState(this._getNewState(newProps))
+  }
+
+  _refresh = () => {
+    this.setState({
+      refreshing: true
+    })
+    this.props.onRefreshDiscover()

```

Comment:
Should add a catch here and display an alert or something that there is no internet connection